### ISSUE TITLE
Crash in Azure Core HTTP Client

### ISSUE THREAD

**Issue by JinmingHu-MSFT** [Azure/azure-sdk-for-cpp#248]
I'm using the latest commit and didn't make any changes to Azure core code.

it crashed in `multi.c`

```
bool Curl_is_in_callback(struct Curl_easy *easy)
{
  return ((easy->multi && easy->multi->in_callback) ||
          (easy->multi_easy && easy->multi_easy->in_callback));
}
```

Exception thrown: read access violation.
easy->**multi** was 0xFFFFFFFFFFFFFE87. occurred

Callstack:
```
>	libcurl-d.dll!Curl_is_in_callback(Curl_easy * easy) Line 3306	C
 	libcurl-d.dll!curl_easy_recv(Curl_easy * data, void * buffer, unsigned __int64 buflen, unsigned __int64 * n) Line 1079	C
 	azure-storage-test.exe!Azure::Core::Http::CurlSession::ReadSocketToBuffer(unsigned char * buffer, __int64 bufferSize) Line 422	C++
 	azure-storage-test.exe!Azure::Core::Http::CurlSession::Read(Azure::Core::Context & context, unsigned char * buffer, __int64 count) Line 386	C++
 	azure-storage-test.exe!Azure::Core::Http::BodyStream::ReadToCount(Azure::Core::Context & context, Azure::Core::Http::BodyStream & body, unsigned char * buffer, __int64 count) Line 32	C++
 	azure-storage-test.exe!Azure::Core::Http::BodyStream::ReadToEnd(Azure::Core::Context & context, Azure::Core::Http::BodyStream & body) Line 51	C++
 	azure-storage-test.exe!Azure::Storage::Blobs::BlobRestClient::Service::ListBlobContainersParseResponse(Azure::Core::Context context, std::unique_ptr<Azure::Core::Http::Response,std::default_delete<Azure::Core::Http::Response>> pHttpResponse) Line 1112	C++
 	azure-storage-test.exe!Azure::Storage::Blobs::BlobRestClient::Service::ListBlobContainers(Azure::Core::Context context, Azure::Core::Http::HttpPipeline & pipeline, const std::string & url, const Azure::Storage::Blobs::BlobRestClient::Service::ListBlobContainersOptions & options) Line 1145	C++
 	azure-storage-test.exe!Azure::Storage::Blobs::BlobServiceClient::ListBlobContainersSegment(const Azure::Storage::Blobs::ListBlobContainersOptions & options) Line 122	C++
 	azure-storage-test.exe!Azure::Storage::Test::BlobServiceClientTest_ListContainers_Test::TestBody() Line 51	C++
 	azure-storage-test.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2599	C++
 	azure-storage-test.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2650	C++
 	azure-storage-test.exe!testing::Test::Run() Line 2696	C++
 	azure-storage-test.exe!testing::TestInfo::Run() Line 2869	C++
 	azure-storage-test.exe!testing::TestSuite::Run() Line 3021	C++
 	azure-storage-test.exe!testing::internal::UnitTestImpl::RunAllTests() Line 5731	C++
 	azure-storage-test.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)() method, const char * location) Line 2599	C++
 	azure-storage-test.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)() method, const char * location) Line 2650	C++
 	azure-storage-test.exe!testing::UnitTest::Run() Line 5313	C++
 	azure-storage-test.exe!RUN_ALL_TESTS() Line 2486	C++
 	azure-storage-test.exe!main(int argc, char * * argv) Line 10	C++
 	[External Code]	
```

It was performing a list containers operation, the response of which is a chunked response (I'm not implying this is an issue specific to chunked response, it may also happen to other HTTP responses).

Seems this issue only exists on Windows, I didn't observe it on Linux yet.

**Comment by RickWinter** [comment]
@JinmingHu-MSFT please add the repro code and repro instructions to the issue.  Reassign to @vhvb1989 after adding the repro instuctions

**Comment by JinmingHu-MSFT** [comment]
Here's repro code

```
// Copyright (c) Microsoft Corporation. All rights reserved.
// SPDX-License-Identifier: MIT

#include "http/pipeline.hpp"

#include <array>
#include <http/curl/curl.hpp>
#include <http/http.hpp>
#include <iostream>
#include <memory>
#include <vector>

using namespace Azure::Core;
using namespace Azure::Core::Http;
using namespace std;

int main()
{
  try
  {
    // Create the Transport
    std::shared_ptr<HttpTransport> transport = std::make_unique<CurlTransport>();

    std::vector<std::unique_ptr<HttpPolicy>> policies;

    // Add the transport policy
    policies.push_back(std::make_unique<TransportPolicy>(std::move(transport)));

    auto httpPipeline = Http::HttpPipeline(policies);

    auto context = Context();

    string host("http://anglesharp.azurewebsites.net/Chunked");

    auto request = Http::Request(Http::HttpMethod::Get, host);

    auto response = httpPipeline.Send(context, request);
    auto response_bodystream = response->GetBodyStream();
    auto response_body = BodyStream::ReadToEnd(context, *response_bodystream);
  }
  catch (Http::CouldNotResolveHostException& e)
  {
    cout << e.what() << endl;
  }
  catch (Http::TransportException& e)
  {
    cout << e.what() << endl;
  }

  return 0;
}
```

**Comment by JinmingHu-MSFT** [comment]
crash stack


```
>	libcurl-d.dll!Curl_is_in_callback(Curl_easy * easy) Line 3306	C
 	libcurl-d.dll!curl_easy_recv(Curl_easy * data, void * buffer, unsigned __int64 buflen, unsigned __int64 * n) Line 1079	C
 	azure_core_with_curl_stream.exe!Azure::Core::Http::CurlSession::ReadSocketToBuffer(unsigned char * buffer, __int64 bufferSize) Line 422	C++
 	azure_core_with_curl_stream.exe!Azure::Core::Http::CurlSession::Read(Azure::Core::Context & context, unsigned char * buffer, __int64 count) Line 386	C++
 	azure_core_with_curl_stream.exe!Azure::Core::Http::BodyStream::ReadToCount(Azure::Core::Context & context, Azure::Core::Http::BodyStream & body, unsigned char * buffer, __int64 count) Line 32	C++
 	azure_core_with_curl_stream.exe!Azure::Core::Http::BodyStream::ReadToEnd(Azure::Core::Context & context, Azure::Core::Http::BodyStream & body) Line 51	C++
 	azure_core_with_curl_stream.exe!main() Line 40	C++
 	[External Code]	
```

**Comment by vhvb1989** [comment]
will be Fixed with #256
